virtme/commands/run.py

```diff
@@ -1840,6 +1874,24 @@ def get_net_mac(index):
     # Load a normal kernel
     qemuargs.extend(["-kernel", kernel.kimg])
     if kernelargs:
+        if args.systemd:
+            os.makedirs(CACHE_DIR, exist_ok=True)
+            with open(SERIAL_GETTY_FILE, "w", encoding="utf-8") as f:
+                f.write("""[Service]
+StandardInput=tty
+StandardOutput=tty
+TTYPath=/dev/%I
+TTYReset=yes
+TTYVHangup=yes
+Environment=""")
+                for arg in kernelargs:
+                    match = re.match(r"(virtme_.*)=(.*)", arg)
+                    if match:
+                        f.write(f"{match.group(1)}='{match.group(2)}' ")
```

Comment:
```suggestion
        if args.systemd:
              init_environment_vars = []
              for arg in kernelargs:
                    match = re.match(r"(virtme_.*)=(.*)", arg)
                    if match:
                        init_environment_vars.append(f"{match.group(1)}={shlex.quote(match.group(2))})
             os.makedirs(CACHE_DIR, exist_ok=True)
             with open(SERIAL_GETTY_FILE, "w", encoding="utf-8") as f:
                 f.write(f"""[Service]
StandardInput=tty
StandardOutput=tty
TTYPath=/dev/%I
TTYReset=yes
TTYVHangup=yes
Environment={shlex.quote(shlex.join(init_environment_vars))}""")
```

(untested, and probably is the indentation off).

Reply:
Thanks, fixed in https://github.com/arighi/virtme-ng/pull/280/commits/918d2069374938b3936c6fbe60d8b5bda72a7aa6. However, the second quote() in Environment= should not be needed, right? Otherwise it yields some weird string that breaks the unit file.

Reply:
For reference,

```
Environment='virtme_hostname=virtme-ng virtme_systemd=1 virtme_root_mods=1 virtme_rw_overlay0=/etc virtme_rw_overlay1=/lib virtme_rw_overlay2=/home virtme_rw_overlay3=/opt virtme_rw_overlay4=/srv virtme_rw_overlay5=/usr virtme_rw_overlay6=/var virtme_rw_overlay7=/tmp virtme_console=ttyS0 '"'"'virtme_stty_con='"'"'"'"'"'"'"'"'rows 76 cols 108 iutf8'"'"'"'"'"'"'"'"''"'"' virtme_chdir=mnt/ext/src/linux/virtme-ng/sys    temd virtme_root_user=1' 
```
vs
```
Environment=virtme_hostname=virtme-ng virtme_systemd=1 virtme_root_mods=1 virtme_rw_overlay0=/etc virtme_rw_overlay1=/lib virtme_rw_overlay2=/home virtme_rw_overlay3=/opt virtme_rw_overlay4=/srv virtme_rw_overlay5=/usr virtme_rw_overlay6=/var virtme_rw_overlay7=/tmp virtme_console=ttyS0 virtme_stty_con='rows 76 cols 108 iutf8' virtme_chdir=mnt/ext/src/linux/virtme-ng/systemd virtme_root_user=1
```

Reply:
I've updated my original answer [1].

[1] https://github.com/arighi/virtme-ng/pull/280#discussion_r2086860872

Reply:
Perfect, it works. Thanks for looking into this. Added in https://github.com/arighi/virtme-ng/pull/280/commits/3988d3ce5731627b2bf8663aeca55e653fdd710d